cryptography/src/bls12381/dkg/types.rs

```diff
@@ -0,0 +1,232 @@
+//!
+//! [Dealer]: crate::bls12381::dkg::Dealer
+//! [Player]: crate::bls12381::dkg::Player
+
+use crate::{
+    bls12381::primitives::{group, poly::Public, variant::Variant},
+    PublicKey, Signature, Signer,
+};
+use bytes::{Buf, BufMut};
+use commonware_codec::{varint::UInt, EncodeSize, FixedSize, Read, ReadExt, Write};
+use commonware_utils::quorum;
+
+/// The signature namespace for DKG share acknowledgements.
+pub const ACK_NAMESPACE: &[u8] = b"DKG_ACK";
+
+/// A [Share] sent by a [Dealer] node to a [Player] node.
+///
+/// Contains the [Dealer]'s public commitment to their polynomial and the specific
+/// share calculated for the receiving [Player].
+///
+/// [Share]: group::Share
+/// [Dealer]: crate::bls12381::dkg::Dealer
+/// [Player]: crate::bls12381::dkg::Player
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Share<V: Variant> {
+    /// The [Dealer]'s public commitment (coefficients of the polynomial).
+    ///
+    /// [Dealer]: crate::bls12381::dkg::Dealer
+    pub commitment: Public<V>,
+    /// The secret share evaluated for the recipient [Player].
+    ///
+    /// [Player]: crate::bls12381::dkg::Player
+    pub share: group::Share,
+}
+
+impl<V: Variant> Share<V> {
+    /// Create a new [Share] message.
+    pub fn new(commitment: Public<V>, share: group::Share) -> Self {
+        Self { commitment, share }
+    }
+}
+
+impl<V: Variant> Write for Share<V> {
+    fn write(&self, buf: &mut impl BufMut) {
+        self.commitment.write(buf);
+        self.share.write(buf);
+    }
+}
+
+impl<V: Variant> EncodeSize for Share<V> {
+    fn encode_size(&self) -> usize {
+        self.commitment.encode_size() + self.share.encode_size()
+    }
+}
+
+impl<V: Variant> Read for Share<V> {
+    type Cfg = u32;
+
+    fn read_cfg(buf: &mut impl Buf, t: &u32) -> Result<Self, commonware_codec::Error> {
+        let q = quorum(*t);
+        Ok(Self {
+            commitment: Public::<V>::read_cfg(buf, &(q as usize))?,
+            share: group::Share::read(buf)?,
+        })
+    }
+}
+
+/// Acknowledgement message sent by a [Player] node back to the [Dealer] node.
+///
+/// Acknowledges the receipt and verification of a [Share] message.
+/// Includes a signature to authenticate the acknowledgment.
+///
+/// [Dealer]: crate::bls12381::dkg::Dealer
+/// [Player]: crate::bls12381::dkg::Player
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Ack<S: Signature> {
+    /// The public key identifier of the [Player] sending the acknowledgment.
+    ///
+    /// [Player]: crate::bls12381::dkg::Player
+    pub public_key: u32,
+    /// A signature covering the DKG round, dealer ID, and the [Dealer]'s commitment.
+    /// This confirms the player received and validated the correct share.
+    ///
+    /// [Dealer]: crate::bls12381::dkg::Dealer
+    pub signature: S,
+}
+
+impl<S: Signature> Ack<S> {
+    /// Create a new [Ack] message, constructing and signing the payload with the provided [Signer].
+    pub fn new<C, V>(
```

Comment:
like in consensus, we still need to accept namespace here or we can have two instances of DKG collide.